### PR TITLE
[PROG-1230] Graceful shutdown

### PR DESCRIPTION
**Pull request by umutdemirel**
## JIRA
[PROG-1230](https://jira.tapjoy.net/browse/PROG-1230)

## Description
This PR adds graceful shutdown functionality that is compatible with our deployment procedures to the tpe_prebid_service project.
The workflow is as follows:
- When a backend service is initiated, it writes it's process id to a file (<process_id>.pid) in a folder (default is './pids' but that path is hardcoded in `grace-shepherd.sh` script,  therefore it would break the system if another value is assigned)
- When we want to deploy a new version of the service, `grace-shepherd.sh` script is run
  - It first notifies the service by sending `-SIGUSR1` signal to the service.
  - The service should be listening to this signal and gracefully shut itself down by closing all connections and ports securely, and finally raise the signal.
  - Then the scripts kill the process by reading the process id from the file we stored during the initialization of the service (./pids/<process_id>.pid)
  - And then, successfully restart the service without any errors or problem

Listening to the `-SIGTERM` and `-SIGINT` has already been implemented in `tpe_prebid_project` and so graceful shutdown functionality.

The only addition is saving the process id functionality and adding `-SIGUSR1` to the list of signals to be listened to.

Please review: @eric-kansas @alanbrent

## How To Test
Tried to test this in TIAB in the following ways:
- I set up a TIAB and send the executable to the ad_service folder in the TIAB (currently we don't have tpe_prebid_service folder yet). 
- I couldn't update the up-start config, so I tried to hack `graceful-shepherd.sh` to set the command to `./prebid_server` and service to `prebid_server-web`.
- Then I tried to run `stop` and `restart` commands.
- It restarted `tpe_prebid_service` succesfully, but the new process is not being spawned and that part of `graceful-shepherd.sh` script is not run.
- Then I send signals to kill the service `>kill -SIGUSR1 <process_id>`
- That way, I could make the process being respawn

I'm not sure if this is the right way, but at least I tried :)

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -66,6 +67,10 @@ type Configuration struct {
 	PemCertsFile string `mapstructure:"certificates_file"`
 	// Custom headers to handle request timeouts from queueing infrastructure
 	RequestTimeoutHeaders RequestTimeoutHeaders `mapstructure:"request_timeout_headers"`
+
+	// Deployment
+	DeployPIDPath string      `mapstructure:"deploy_pid_path"`
+	DeployPIDMode os.FileMode `mapstructure:"deploy_pid_mode"`
```
[the remark this answers is not in the document]
Comment (umutdemirel) [reply config/config.go:73]: What would be the behavior of `DeployPIDEnabled`? Should we not save pids when `DeployPIDEnabled == false`? Or should we not listen to `SIGUSR1` when `DeployPIDEnabled  == false`? Or do we need to do something in `graceshepherd.sh` file?

File: grace-shepherd.sh [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+            log "Deleting ${COMMAND} pid file with no running process '$PID_FILE'"
+            rm $PID_FILE 2> /dev/null  || log "Failed to delete pid file '$PID_FILE': $!"
+        fi
+    done
+    if $multi_master ; then
+        log "Found more than one not old ${COMMAND} process running. Pids are '$pid $extra_pids'."
+        log "Killing them all and restarting."
+        kill -9 $pid $extra_pids
+        exit 1
+    fi
```
Comment (umutdemirel) [inline grace-shepherd.sh:103]: @alanbrent This part of the script handles other pids. It gets the folder and clean all pid files.

File: config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alanbrent) [reply config/config.go:73]: We should not create the PIDs in that case. You can still listen to `SIGUSR1` if you like, although it wouldn't be strictly necessary.

File: grace-shepherd.sh (hunk 2 above)
Comment (alanbrent) [reply grace-shepherd.sh:103]: Ah yes I remember this now. It makes enough sense for production purposes. I have run into issues with PIDs in localdev, and now I think I've learned that it's because we don't use `grace-shepherd.sh` in localdev because it's tightly coupled with an assumed `init` process managing the OS.

TLDR, if you end up implementing that `DEPLOY_PID_ENABLED` environment variable we can just leave it at `false` (the appropriate default in my _personal_ opinion) and we would never have to deal with them locally.

**Comment by umutdemirel** [comment]
@alanbrent 
Thanks for your comments. I added `DeployPIDEnabled` field in configs with the default value of `false`. Now, we're able to control the storing of `pid` files. One important note is that the corresponding env var is `PBS_DEPLOY_PID_ENABLED`.
It would be awesome if you can review one more round.
